path_planning.py

```diff
@@ -0,0 +1,321 @@
+        for neighbor, neighbor_cost in neighbors:
+            if(neighbor in  best_costs):
+                #print("found=" , neighbor)     
+                ncost,npoint = best_costs.get(neighbor)
+                if(ncost > cost + neighbor_cost) :        
+                    #print("update cost for ", neighbor ,"newcost", cost + neighbor_cost)
+                    best_costs[neighbor] = ( cost + neighbor_cost ,  node )   
+                    heapq.heappush (open_list,  (cost + neighbor_cost,  (neighbor) ) )     
+            else:    
+                    #print(neighbor ,"newcost", cost + neighbor_cost)
+                    best_costs[neighbor] = ( cost + neighbor_cost ,  node )    
+                    heapq.heappush (open_list,  (cost + neighbor_cost,  (neighbor) ) )       
+                    
+    if(found == False) :                   
+        return (visit_count, -1, None)
+    
+   
+    node = goal
+    bestcost,point = best_costs.get(node)   
+    retpath: List[Point] = []
+    while(1):       
+        cost,point = best_costs.get(node)          
+        retpath.insert(0,node)
+        if(compare_points(point,start)) :
+            retpath.insert(0,point)
+            break;
+        node = point
+     
+    print("bestcost" ,bestcost    )
+    print("path" ,retpath    )
+   
+ 
+    return (visit_count, bestcost, retpath)    
+
+def breadth_first(terrain:Map, start:Point, goal:Point) -> Tuple[int, int, Optional[List[Point]]]:
+    open_list: List[Point] = [start]   
+ 
+    # We'll treat start specially
+    best_costs: Dict[Point, Tuple[int, Point]] = {start:(0, start)}
+    visit_count = 0
+    found_count = 0
+    while open_list:
+        # Breadth-first search takes the first thing from the list...
+        node = open_list.pop(0)
+       
+        cost,point = best_costs.get(node)     
+        #print("node=",node,"cost=",cost)
+        
+        if(compare_points(node,goal)):
+            #print ("found it") 
+            found_count += 1;
+            continue
+        
+        visit_count += 1
+        
+        neighbors = find_neighbors(terrain, node)
+        #print(neighbors)      
+        for neighbor, neighbor_cost in neighbors:   
+            if(neighbor in  best_costs):
+                #print("found=" , neighbor)     
+                ncost,npoint = best_costs.get(neighbor)
+                if(ncost > cost + neighbor_cost) :        
+                    #print("update cost for ", neighbor ,"newcost", cost + neighbor_cost)
+                    best_costs[neighbor] = ( cost + neighbor_cost ,  node )   
+                    open_list.insert( len(open_list)  , (neighbor) )      
+            else:    
+                    #print(neighbor ,"newcost", cost + neighbor_cost)
+                    best_costs[neighbor] = ( cost + neighbor_cost ,  node )    
+                    open_list.insert( len(open_list)  , (neighbor) )                                                          

```

Comment:
Might be cleaner to handle these cases using an `or` and  `not in` since you do the same thing when the neighbor isn't in `best_costs` and when a cheaper cost has been found.